### DISCUSSION TITLE
Storage texture and textureLoad

### DISCUSSION THREAD

**Discussion by ottoville** [gpuweb/gpuweb#4121] [Q&A] [answered]
Hello,

In a wgsl specs it is described that a storage texture is a texture that "supports accessing a single texel without the use of a sampler."

However, the textureLoad function that is used to load single texel without sampler, does not have overload with texture_storage_2d type of texture, so with that type it is not possible to access a single texel using textureLoad.

How is the storage texture supposted to be used?

**Comment by tlecoz** [comment]
I'm not sure to understand the question,  but in the library I'm working on , I use the computePass.dispatchWorkgroup to store the dimension of the picture like that 

`computePass.dispatchWorkgroups(image.width,image.height,1);`

then, in the computeShader, I can use workgroup.id as pixel position  like that 

```
let id = vec2<i32>(workgroupId.xy);
let dim = vec2<f32>(textureDimensions(image));
var col = textureSampleLevel(image, textureSampler, (0.5+vec2<f32>(id)) / dim,0.0);
```

the entire code of my shader : 

```
struct Uniforms {
    time:f32,
}

@binding(0) @group(0) var<uniform> uniforms:Uniforms;
@binding(1) @group(0) var textureSampler:sampler;
@binding(0) @group(1) var image:texture_2d<f32>;
@binding(1) @group(1) var image_out : texture_storage_2d<rgba8unorm, write>;

@compute @workgroup_size(1,1,1)
fn main(@builtin(workgroup_id) workgroupId:vec3<u32>) {
   var time:f32 = uniforms.time;
   let id = vec2<i32>(workgroupId.xy);
   let dim = vec2<f32>(textureDimensions(image));
   var col = textureSampleLevel(image, textureSampler, (0.5+vec2<f32>(id)) / dim,0.0);

   if(floor(time) % 2 == 0) { 
       col.r = 0.5 + sin( time) * 0.5;
       col.g = 0.5 + cos( time + col.b) * 0.5;
   }
   
   textureStore(image_out, id, col);
   
}
```

Hope it help ;)

**Comment by ottoville** [comment]
Thank you for the answer.
So storage textures are type of textures that can be written into with compute shader, but not read? If so, why do they have a read/write access mode?

**Comment by tlecoz** [comment]
I'm not sure that you can diretcly read/write a texture (or a vertexBuffer) ,that's not how I use it at least. 
I'm not an expert at all with computeShader , I may be totally wrong. 
I followed the logic behind the webgpu samples : https://webgpu.github.io/webgpu-samples/

they use what they call a "ping-pong" structure. 

you have 2 buffers A and B , this is just buffers , it can be read or write , it depends on the bufferUsage defined in the bindgroup that will exploit it. 

you have 2 bindgroups 

in the first one, bufferA has an accessMode set to "read" ad bufferB has an accessMode set to "write". 
in the second bindgroup, you have the bufferA set to "write" and bufferB set to "read" 

Then, at the end of each frame, you swap the bindgroup 

So at the end of the shader you write inside BufferB  , then you swap the bindgroup, so BufferB become the buffer defined to be "read" and bufferA become the buffer defined to store the data, etc... 

it's a bit tricky at first sight but totally doable :) 

EDIT : it works exactly the same way with VertexBuffers

**Comment by Kangz** [comment, marked as answer]
WebGPU doesn't allow reading from storage textures yet. The reason is that some Metal devices can't support this capability and wouldn't be able to run WebGPU if readable storage textures were a core feature. However it is in the list of short-term extensions to the API that are being considered. See #1772

**Comment by tlecoz** [comment]
"it is in the list of short-term extensions to the API that are being considered."
What are the other short-term extension to the API that are being considered ? :)

**Reply by Kangz** [reply]
There isn't a pinned down list yet, but things the members of the group are thinking of are listed in the [Milestone 2](https://github.com/gpuweb/gpuweb/issues?q=is%3Aopen+is%3Aissue+milestone%3A%22Milestone+2%22
) and [Polish Post-V1](https://github.com/gpuweb/gpuweb/issues?q=is%3Aopen+is%3Aissue+milestone%3A%22Polish+post-V1%22) issues.

**Reply by tlecoz** [reply]
thank you !

**Comment by ottoville** [comment]
Thank you for the answer.
I know this is a wrong place to post this but as I know there are chromium developers it is for me quickest way to get help.
When creating a storage texture with size of 8192*8192 (maximum allowed size) and format rgba8unorm, there will be error in chrome
```
Buffer size (268435460) exceeds the max buffer size limit (268435456).
    at ValidateBufferDescriptor (..\..\third_party\dawn\src\dawn\native\Buffer.cpp:139)
    at CreateBuffer (..\..\third_party\dawn\src\dawn\native\Device.cpp:1533)
    at AllocateInternal (..\..\third_party\dawn\src\dawn\native\DynamicUploader.cpp:46)
    at Allocate (..\..\third_party\dawn\src\dawn\native\DynamicUploader.cpp:136)
    at ClearTexture (..\..\third_party\dawn\src\dawn\native\d3d12\TextureD3D12.cpp:816)
    at EnsureSubresourceContentInitialized (..\..\third_party\dawn\src\dawn\native\d3d12\TextureD3D12.cpp:871)
    at operator() (..\..\third_party\dawn\src\dawn\native\d3d12\CommandBufferD3D12.cpp:362)
    at Iterate (..\..\third_party\dawn\src\dawn\native\SubresourceStorage.h:385)
    at TransitionAndClearForSyncScope (..\..\third_party\dawn\src\dawn\native\d3d12\CommandBufferD3D12.cpp:366)
    at RecordComputePass (..\..\third_party\dawn\src\dawn\native\d3d12\CommandBufferD3D12.cpp:1188)
    at RecordCommands (..\..\third_party\dawn\src\dawn\native\d3d12\CommandBufferD3D12.cpp:776)
    at SubmitImpl (..\..\third_party\dawn\src\dawn\native\d3d12\QueueD3D12.cpp:52)


```
8192 x 8192 x 4 = 268435456 so there is an extra word that makes the buffer too large. This is a bug in chrome I assume?

**Reply by kainino0x** [reply]
FYI questions like this are best asked on https://matrix.to/#/#webgpu-dawn:matrix.org or https://groups.google.com/g/dawn-graphics

In this case it does sound like a bug. Could you please file a bug (with repro case if possible) on https://crbug.com/dawn ?
